src/CodeGen_Hexagon.cpp

```diff
@@ -535,92 +540,202 @@ Value *CodeGen_Hexagon::call_intrin_cast(llvm::Type *ret_ty,
-        int bytes_off = start * (element_bits / 8);
-        int reverse_bytes = (native_vector_bits() / 8) - bytes_off;
-        Intrinsic::ID intrin_id = IPICK(Intrinsic::hexagon_V6_valignbi);
-        if (reverse_bytes <= 7) {
-            intrin_id = IPICK(Intrinsic::hexagon_V6_vlalignbi);
-            bytes_off = reverse_bytes;
-        }
-        Value *hi, *lo;
-        if (is_vcombine) {
-            hi = call->getArgOperand(0);
-            lo = call->getArgOperand(1);
-        } else {
-            hi = call_intrin_cast(ret_ty, IPICK(Intrinsic::hexagon_V6_hi), {vec});
-            lo = call_intrin_cast(ret_ty, IPICK(Intrinsic::hexagon_V6_lo), {vec});
+    }
+
+    // Try to rewrite shuffles of (maybe strided) ramps.
+    bool is_strided_ramp = true;
+    int stride = indices.size() > 1 ? indices[1] - indices[0] : 1;
+    for (int i = 1; i + 1 < static_cast<int>(indices.size()); i++) {
+        if (indices[i] + stride != indices[i + 1]) {
+            is_strided_ramp = false;
+            break;
         }
-        Value *b = codegen(bytes_off);
-        return call_intrin_cast(ret_ty, intrin_id, {hi, lo, b});
     }
-    return CodeGen_Posix::slice_vector(vec, start, size);
-}
 
-Value *CodeGen_Hexagon::concat_vectors(const vector<Value *> &v) {
-    bool B128 = target.has_feature(Halide::Target::HVX_128);
+    if (!is_strided_ramp) {
+        // This is not a strided ramp, let LLVM handle it.
+        return CodeGen_Posix::shuffle_vectors(a, b, indices);
+    }
 
-    if (v.size() == 2 && v[0]->getType() == v[1]->getType()) {
-        llvm::Type *v_ty = v[0]->getType();
-        int vec_elements = v_ty->getVectorNumElements();
-        int element_bits = v_ty->getScalarSizeInBits();
-        if (vec_elements*element_bits == native_vector_bits()) {
-            return call_intrin_cast(llvm::VectorType::get(v_ty->getScalarType(), vec_elements*2),
-                                    IPICK(Intrinsic::hexagon_V6_vcombine),
-                                    {v[1], v[0]});
+    int start = indices[0];
+
+    if (stride == 1) {
+        if (result_ty == native2_ty && a_ty == native_ty && b_ty == native_ty) {
+            // This is a concatenation of a and b, where a and b are
+            // native vectors. Use vcombine.
+            internal_assert(start == 0);
+            return call_intrin_cast(native2_ty, IPICK(Intrinsic::hexagon_V6_vcombine), {b, a});
+        }
+        if (result_ty == native_ty && a_ty == native2_ty && max < a_elements) {
+            // Extract a and b from a double vector.
+            b = call_intrin_cast(native_ty, IPICK(Intrinsic::hexagon_V6_hi), {a});
+            a = call_intrin_cast(native_ty, IPICK(Intrinsic::hexagon_V6_lo), {a});
+            a_ty = a->getType();
+            b_ty = b->getType();
+            a_elements = a_ty->getVectorNumElements();
+            b_elements = b_ty->getVectorNumElements();
+        }
+        if (start == 0 && result_ty == a_ty) {
+            return a;
```

Comment:
Consider a shuffle of the type 
shuffle_vectors(dbl_vector_a_b, 0, 31, 1, 32, 2, 33..., 31,63)  
It is a weird interleaving operation that is just a contrived example that won't be equivalent to returning a. If you think this is far too contrived a case to appear in the real world, then please ignore this comment.